### PR TITLE
[Network] Improve application client and service APIs.

### PR DESCRIPTION
**Pull request by JoshLind**
Note: most of this PR is just updating tests to handle the new simplified network interfaces.

### Description

This PR follows on from https://github.com/aptos-labs/aptos-core/pull/6013 and continues to clean up the network sending and receiving APIs used by the various applications (e.g., consensus, mempool and state sync). Specifically, the PR:
1. Introduces a new struct `NetworkServiceEvents` which is passed directly to all applications to allow them to process events from the network, including the network ID on which the event was received (e.g., validator, vfn, or public).
2. Updates the `aptos-node` crate to only expose the `NetworkClient` and `NetworkServiceEvents` APIs to each application. This simplifies the number of structs/items being passed around the codebase to each application and makes the code cleaner.
3. Unifies all application configurations (e.g., direct send protocols, RPC protocols, network handlers, etc.) into a single file (`network.rs`) used by the `aptos-core` crate. Now, we now longer need to duplicate application configs and leak network details across the codebase. Instead, each application is given a `NetworkClient` and `NetworkServiceEvents` API, where the network registration has already been done. This provides better code separation.
4. As a result of unifying the network events across multiple network IDs, there is now only a single storage service running on each node, that handles all storage requests across the networks. This is in contrast to the old network code, where a storage service was started for each network ID, which was unneccessary and sub-optimal.
5. We clean up the `NetworkApplicationConfig` struct so that it now provides simple separation between the client and service side, making it easier to reason about protocol IDs and preferences (as required by consensus).
6. By moving all network configuration definitions into a single location, and removing the code coupling, we can now make the applications more configurable, e.g., we can update consensus to disable compression using a simple config variable! This wasn't previously possible without a huge refactor ('cc @zekun000).
7. Various other code cleanups and removal of boilerplate.

Once this lands we should now have a simpler abstraction between the applications and networking code! 😄 

### Test Plan
Existing test infrastructure.

### CODE REVIEW

File: mempool/src/shared_mempool/network.rs [hunk 1]
```diff
@@ -63,24 +60,6 @@ pub enum MempoolSyncMsg {
...
-/// Returns a network application config for the mempool client and service
-pub fn mempool_network_config(max_broadcasts_per_peer: usize) -> NetworkApplicationConfig {
-    NetworkApplicationConfig::client_and_service(
-        [ProtocolId::MempoolDirectSend],
-        aptos_channel::Config::new(max_broadcasts_per_peer)
-            .queue_style(QueueStyle::KLAST)
-            .counters(&counters::PENDING_MEMPOOL_NETWORK_EVENTS),
-    )
-}
-
```
Comment (davidiw) [inline mempool/src/shared_mempool/network.rs:83]: so we move more data away from the application that uses networking and more into the node... hmm...

**Review by davidiw — APPROVED** [review]
I'd like to better understand the principle in organization. I feel mixed about seeing more code heading away from their respective endpoints and into the node with respect to configuration.... but there's a lot of decent clean up here and we need to iterate more.

**Comment by JoshLind** [comment]
Thanks @davidiw!

> I'd like to better understand the principle in organization. 

The current approach is to remove as many leaky abstractions as possible, provide cleaner interfaces between the networking stack and the AptosNet applications, and implement better encapsulation (generally). There is a high-level of coupling at the moment, so I'm partitioning things first, so it's easier to manage and change them in the future.

> I feel mixed about seeing more code heading away from their respective endpoints and into the node with respect to configuration

I see what you're saying. Personally, I think of `aptos-core` as the crate that simply wires together a node (i.e., using configs, components, runtimes, etc.). As part of that, the network application configs should also be instantiated. With this, each AptosNet application now only depends on a single sending and receiving interface with all the dirty configuration and instantiation already done. The benefit is that: (i) there's now a single place to locate all network wiring; (ii) applications don't have to be tightly coupled to the current network implementation and channel choice; and (iii) it's also easier to update/compare network configs (e.g., I've now seen that mempool uses a KLAST channel, while everything else uses FIFO. Which I think is odd...)

With that said, once most of the cleanups and refactoring is done, we could always move them back into each application (if we want to). I'm not super worried about the network configs, specifically. I just want to partition and clean things up for now :) Happy to migrate things once everything is clean.

File: mempool/src/shared_mempool/network.rs (hunk 1 above)
Comment (JoshLind) [reply mempool/src/shared_mempool/network.rs:83]: See my response below 😸

File: network/src/protocols/network/mod.rs [hunk 2]
```diff
@@ -79,51 +76,68 @@ impl<TMessage: PartialEq> PartialEq for Event<TMessage> {
...
+            rpc_protocols_and_preferences,
         }
     }
+}
+
+/// Configuration needed for the service side of AptosNet applications
+#[derive(Clone)]
+pub struct NetworkServiceConfig {
+    /// Direct send protocols for the application (sorted by preference, highest to lowest)
+    pub direct_send_protocols_and_preferences: Vec<ProtocolId>,
```
Comment (sitalkedia) [inline network/src/protocols/network/mod.rs:104]: Hmm.. with this change, we need to duplicate protocol_ids both on the client and server configuration - is my understanding correct? Can you explain what happens if there is a configuration mismatch?

File: aptos-node/src/network.rs [hunk 3]
```diff
@@ -1,39 +1,112 @@
...
 use aptos_time_service::TimeService;
 use aptos_types::chain_id::ChainId;
+use serde::{Deserialize, Serialize};
 use std::{collections::HashMap, sync::Arc};
 use tokio::runtime::Runtime;
 
-const MEMPOOL_NETWORK_CHANNEL_BUFFER_SIZE: usize = 1_024;
+// TODO: should the applications not make these configurable?
+// Application network channel sizes
+const CONSENSUS_NETWORK_CHANNEL_BUFFER_SIZE: usize = 1024;
```
[the remark this answers is not in the document]
Comment (JoshLind) [reply aptos-node/src/network.rs:34]: Yep, same size. It was actually previously this (https://github.com/aptos-labs/aptos-core/blob/main/network/src/constants.rs#L19) which is marked as test only code 😆

File: network/src/protocols/network/mod.rs (hunk 2 above)
Comment (JoshLind) [reply network/src/protocols/network/mod.rs:104]: Yes, but I think this is okay, for a couple reasons: (i) protocols are actually negotiated between client and server, e.g., there is an order preference that is done when messages are actually sent. See here (https://github.com/aptos-labs/aptos-core/blob/main/network/src/application/interface.rs#L112); (ii) we only wire this up once, using the same constants inside each ApplicationNetworkConfig, so there shouldn't be a case where they disagree within a single node. We can always add some tests around this (to be extra sure no-one misconfigures a node); and (iii) I can see future use cases where we actually want this to be the case (e.g., to maintain backward compatibility when rolling out new services, etc.)

**Review by sitalkedia — APPROVED** [review]
Thanks for clean-up - the networking interfaces look in a much better shape now.

**Pull request merged**